go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,259 @@
+	// and can be built based on the combination of all the other ids.
+	PrimaryIDKey = types.OvnK8sPrefix + "/id"
+
+	// Owner Object Types
+	EgressFirewallDNSOwnerType = "EgressFirewallDNS"
+	EgressQoSOwnerType         = "EgressQoS"
+	NetworkPolicyOwnerType     = "NetworkPolicy"
+	NamespaceOwnerType         = "Namespace"
+	HybridNodeRouteOwnerType   = "HybridNodeRoute"
+)
+
+// DbIndex is a structure representing a set of db object ExternalIDs, used to uniquely identify
+// an object in the db (as a primary index), and to filter out a subset of objects based on some IDs
+// (as a secondary index).
+// DbIndex consists of 2 parts:
+// - OwnerController and OwnerObjectType define who manages given object. It is required to communicate with the other
+// entities that create/update/delete db objects.
+// - ObjectIDs uniquely identify objects within the same OwnerController and OwnerObjectType.
+// e.g. if there is only 1 object owned by given controller and objectType in the system, ObjectIDs be empty.
+//
+// Let's consider address set db object that is uniquely identified by namespace and name. It is managed by
+// "controller-1" and that controller uses objectType "AddressSetType" for those objects. We can define
+// addrSetIDs struct to work with these ids, for that it needs to implement DbObjectIDs interface.
+//
+//	const (
+//		ipFamilyKey = "ipFamily"
+//		Controller = "controller-1"
+//		ObjectType = "AddressSetType"
+//	)
+//
+//	type addrSetIDs struct {
+//		NamespaceName string
+//      ipFamily string
+//	}
+//
+//	func (a *addrSetIDs) GetExternalIDs() map[string]string {
+//		return map[string]string{
+//			SecondaryIndexKey: a.NamespaceName,
+//			ipFamilyKey:      a.ipFamily,
+//		}
+//	}
+//	func (a *addrSetIDs) GetUniqueID() string {
+//		uniqueID := a.NamespaceName
+//		if a.ipFamily != "" {
+//			uniqueID += ":" + a.ipFamily
+//		}
+//		return uniqueID
+//	}
+//
+// This struct may be used to create DbIndex
+//	dbIndex := NewDbIndex[*addrSetIDs](Controller, ObjectType, &addrSetIDs{
+//		NamespaceName: "namespace1_name1",
+//		ipFamily: "ipv4",
+//	})
+//
+// This index may be used to create db object
+//	as := &nbdb.AddressSet{
+//		Name:        dbIndex.GetUniqueID(),
+//		ExternalIDs: dbIndex.GetExternalIDs(),
+//	}
+//
+// The following ExternalIDs will be created for this example
+// {
+//		"k8s.ovn.org/controller": "controller-1"
+//		"k8s.ovn.org/object-type": "AddressSetType"
+//		"k8s.ovn.org/name": "namespace1_name1"
+//		"k8s.ovn.org/id": "AddressSetType.controller-1.k8s.ovn.org/namespace1_name1:ipv4"
+//		"AddressSetType.controller-1.k8s.ovn.org/ipFamily": "ipv4"
+// }
+// AddressSet.Name = AddressSetType.controller-1.k8s.ovn.org/namespace1_name1:ipv4
+//
+// Check addressset.DbIDs object for more examples.
+// To ensure correct usage of DbIndex, and field verification, always use NewDbIndex to create the object.
+type DbIndex[T DbObjectIDs] struct {
```

Comment:
Why would we need `DbObjectIDs` as a parameter type instead of just as a value type? I feel like I am missing something.

Reply:
I just wanted to use structs instead of maps, to enable expressions like `dbIndex.ObjectIDs.NamespaceName` 
but maybe that is too complicated and should be reworked

Reply:
Still confused :P
I don't see that type of expression in `DbIndex` itself. Also thought that generics did not allow accessing the fields of a parameter types :thinking: